modules/provider.go

```diff
@@ -128,3 +135,205 @@ func (p *provider) Warning() bool {
 		return false
 	}
 }
+
+// intToHex converts an int64 to a "0x" prefixed hex string
+func intToHex(n int64) string {
+	return fmt.Sprintf("0x%x", n)
+}
+
+func hexToInt(hex string) (int64, error) {
+	// Check for empty string
+	if len(hex) == 0 {
+		return 0, fmt.Errorf("empty hex string")
+	}
+
+	// Check for minimum length (0x + at least one digit)
+	if len(hex) < 3 {
+		return 0, fmt.Errorf("invalid hex format: too short")
+	}
+
+	// Verify 0x prefix
+	if hex[:2] != "0x" {
+		return 0, fmt.Errorf("invalid hex format: missing 0x prefix")
+	}
+
+	hexNum := hex[2:]
+	return strconv.ParseInt(hexNum, 16, 64)
+}
+
+func parseBlockNumber(result interface{}) (int64, error) {
+	switch result := result.(type) {
+	case string:
+		if result == "" || result[:2] != "0x" {
+			return 0, errors.New("Invalid block number")
+		}
+		// Convert the hexadecimal string to an int64
+		blockNumber, err := hexToInt(result)
+		if err != nil {
+			return 0, errors.Wrap(err, "Error converting block number")
+		}
+		return blockNumber, nil
+	case float64:
+		return int64(result), nil
+	default:
+		return 0, errors.New("unsupported block number type")
+	}
+}
+
+func (p *provider) getLatestBlockNumber(hcMethod string) (int64, int, error) {
+	payload := []byte(fmt.Sprintf(`{"jsonrpc":"2.0","method": "%s","params":[],"id":1}`, hcMethod))
+
+	// Send the POST request
+	resBytes, statusCode, err := p.httpClient.Post(p.HttpUrl, p.Headers, []byte(payload), p.AuthClient())
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "Error sending POST request")
+	}
+
+	if *statusCode == http.StatusServiceUnavailable || *statusCode == StatusOriginUnreachable {
+		return 0, *statusCode, errors.New("Network Unavailable")
+	}
+
+	// response struct
+	var respObject map[string]interface{}
+
+	// Unmarshal the response
+	err = json.Unmarshal(resBytes, &respObject)
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "Error unmarshalling response")
+	}
+
+	if _, ok := respObject["result"]; !ok {
+		return 0, 0, errors.New("Error getting block number from response")
+	}
+
+	blockNumber, err := parseBlockNumber(respObject["result"])
+	if err != nil {
+		return 0, 0, err
+	}
+	return blockNumber, *statusCode, nil
+}
+
+// getEarliestBlockNumber gets the earliest block number from the provider
+// First checks block 1, then uses binary search if not found
+// This is only enabled for EVM based networks
+func (p *provider) getEarliestBlockNumber(getBlockNumberMethod string, retryCount int) (int64, int, error) {
```

Comment:
I think you can simplify this by doing the binary search as part of the main function and rewriting it as recursive.

```suggestion
func (p *provider) getEarliestBlockNumber(getBlockNumberMethod string, lastFound, left, right int) (int64, int, error) {

if left >= right {
   return lastFound
}

mid := left + (right-left)/2

// Create the JSON-RPC payload for getting the block
payload := []byte(fmt.Sprintf(`{"jsonrpc":"2.0","method":"%s","params":["%s", false],"id":1}`,
    getBlockNumberMethod, intToHex(mid)))

....
			
if result, ok := respObject["result"]; ok && result != nil {
return    getEarliestBlockNumber(getBlockNumberMethod string, mid, left, mid - 1)
} else {
return    getEarliestBlockNumber(getBlockNumberMethod string, mid, mid = 1, left)
}
....


```